frontend/public/components/secrets/create-secret.tsx

```diff
@@ -167,11 +199,93 @@ class WebHookSecretSubform extends React.Component<WebHookSecretSubformProps, We
   }
 }
 
+class SourceSecretSubform extends React.Component<SourceSecretSubformProps, SourceSecretSubformState> {
+  constructor (props) {
+    super(props);
+    this.state = this.state = {
+      authenticationType: SecretType.basicAuth,
+      stringData: {
+        username: '',
+        password: '',
+        privateKey: '',
+      },
+    };
+    this.changeData = this.changeData.bind(this);
+    this.changeAuthenticationType = this.changeAuthenticationType.bind(this);
+  }
+  changeAuthenticationType(event) {
+    this.setState({
+      authenticationType: event.target.value
+    }, () => this.props.onChange(this.state));
+  }
+  changeData(event) {
+    const updatedData = _.assign(this.state.stringData, {[event.target.name]: event.target.value});
+    this.setState({
+      stringData: updatedData
+    }, () => this.props.onChange(this.state));
+  }
+  render () {
+    const basicAuthSubform = <React.Fragment>
```

Comment:
Do these need to live inside `render`?  I'd prob rather favor smaller components & just pull them out as separate functions:

```JavaScript 
 const basicAuthSubform: React.SFC = ({props}) => <React.Fragment> ... </React.Fragment>
```
I hesitate to risk `render` bloating over time.